File: study/dataStructures/linkedLists/singlyLinkedLists.js
// A Singly linked list contains a set of nodes, and think of a node as blocks. This nodes have to elements, the values of the data you want to store and a pointer to the next node in line. The fist node es called the head and the last node is called the tail. Linked lists are we call null terminated, which signifies that it's the ends of the list.

// linked lists: apple --> grapes --> pears -->

// apples
// 8947 --> grapes
//          8742 --> pears
//                   372 --> null

//https://visualgo.net/en/list?slide=3

// prepend O(1) - Add into the beginning.
// append O(1) - Add at the end.
// lookup O(n) - finding for an item.
// insert O(n)
// delete O(n)

// A pointer is simply a reference.

// let obj1 = { a: null };
// let obj2 = obj1; // it is a pointer. I is nota a copy in memory both objects pointer to the same location in memory.

// obj1.a = 'Hi';

// console.log('1', obj1);
// console.log('2', obj2);

// 10 --> 5 --> 16

// let myLinkedList = {
//    head: {
//       value: 10,
//       next: {
//          value: 5,
//          next: {
//             value: 16,
//             next: null // because of null it is the tail
//          }
//       }
//    }
// }

class ListNode {
   constructor(value) {
      this.value = value,
         this.next = null
   }
}

class LinkedList {
   constructor(value) {
      this.head = new ListNode(value);
      this.tail = this.head;
      this.length = 1;
   }

   append(value) {
      const newNode = new ListNode(value);
      this.tail.next = newNode;
      this.tail = newNode;
      this.length++;
      return this;
   }

   prepend(value) {
      const newNode = new ListNode(value);
      newNode.next = this.head;
      this.head = newNode;
      this.length++;
      return this;
   }

   lookup(value) {
      return this.traverseToKey(value);
   }

   removeByValue(key) {
      let currentNode = this.head;
      let previousNode = this.head;
      while (currentNode.next !== null) {
         if (currentNode.value === key) {
            if (currentNode === previousNode) {
               this.head = currentNode.next;
               this.length--;
               return this.printList();
            } else {
               previousNode.next = currentNode.next;
               this.length--;
               return this.printList();
            }
         }
         previousNode = currentNode;
         currentNode = currentNode.next;
      }
      if (currentNode.next === null && currentNode.value === key) {
         previousNode.next = null;
         this.tail = previousNode;
         this.length--;
         return this.printList();
      }
      return false;
   }

   removeByIndex(index) {
      if (index > this.length - 1)
         return false;
      if (index === 0) {
         const newHeadNode = this.head.next;
         this.head = newHeadNode;
         this.length--;
         return this.printList();
      }
      const previousNode = this.traverseToIndex(index - 1);
      const currentNode = previousNode.next;
      previousNode.next = currentNode.next;
      if (previousNode.next === null)
         this.tail = previousNode;
      this.length--;
      return this.printList();
   }

   insertByIndex(index, value) {
      if (index >= this.length) {
         this.append(value);
         return this.printList();
      }
      if (index === 0) {
         this.prepend(value);
         return this.printList();
      }
      const newNode = new ListNode(value);
      const currentNode = this.traverseToIndex(index);
      const holdingPointer = currentNode.next;
      currentNode.next = newNode;
      newNode.next = holdingPointer;
      this.length++
      return this.printList();
   }

   insertByValue(key, value) {
      let currentNode = this.traverseToKey(key);
      if (currentNode) {
         if (currentNode.next === null) {
            this.append(value);
            return this.printList();
         }
         const newNode = new ListNode(value);
         newNode.next = currentNode.next;
         currentNode.next = newNode;
         this.length++
         return this.printList();
      } else {
         this.append(value);
         return this.printList();
      }
   }

   traverseToIndex(index) {
      let counter = 0;
      let currentNode = this.head;
      while (counter !== index) {
         currentNode = currentNode.next;
         counter++;
      }
      return currentNode;
   }

   traverseToKey(key) {
      let currentNode = this.head;
      while (currentNode.next !== null) {
         if (currentNode.value === key)
            return currentNode;
         currentNode = currentNode.next;
      }
      return false;
   }

   printList() {
      const array = [this.head.value];
      let currentNode = this.head;
      while (currentNode.next !== null) {
         currentNode = currentNode.next;
         array.push(currentNode.value);
      }
      return array;
   }

   reverse() {
      if (this.length === 1)
         return this.printList();
      let currentNode = this.head;
      let reversedLinkedList = new LinkedList(currentNode.value);
      while (currentNode.next !== null) {
         currentNode = currentNode.next;
         reversedLinkedList.prepend(currentNode.value);
      }
      return reversedLinkedList.printList();
   }

   reverse2() {
      if (this.length === 1)
         return this.printList();
      let first = this.head;
      let second = this.head.next;
      this.tail = this.head;
      while (second) {
         const holdNode = second.next;
         second.next = first;
         first = second;
         second = holdNode;
      }
      this.head.next = null;
      this.head = first;
      return this.printList();
   }
};

const myNewLinkedList = new LinkedList(10);
myNewLinkedList.append(5);
myNewLinkedList.append(16);
myNewLinkedList.prepend(1);
console.log(myNewLinkedList.printList());
myNewLinkedList.insertByValue(10, 2);
myNewLinkedList.insertByValue(5, 3);
myNewLinkedList.insertByValue(250, 50);
myNewLinkedList.insertByValue(1, 99);
console.log(myNewLinkedList.insertByIndex(3, 123));
console.log(myNewLinkedList.insertByIndex(200, 140));
console.log('tail:', myNewLinkedList.tail);

myNewLinkedList.insertByValue(140, 141);
console.log('tail:', myNewLinkedList.tail);
console.log(myNewLinkedList.lookup(80));
console.log(myNewLinkedList.lookup(123));
console.log(myNewLinkedList.removeByIndex(3));
console.log(myNewLinkedList.removeByIndex(0));
console.log(myNewLinkedList.head);
console.log(myNewLinkedList.removeByValue(3));
console.log(myNewLinkedList.removeByValue(1));
console.log(myNewLinkedList.lookup(3));
console.log(myNewLinkedList.removeByValue(141));
console.log('tail:', myNewLinkedList.tail);
console.log(myNewLinkedList.removeByIndex(6));
//console.log('tail:', myNewLinkedList.tail);
console.log(myNewLinkedList.reverse())
console.log(myNewLinkedList.reverse2())

module.exports = LinkedList;

File: Python/data_structures/singly_linked_lists.py
class ListNode:
    def __init__(self, value):
        self.val = value
        self.next = None


class LinkedList:
    def __init__(self, value):
        if type(value) == list:
            self.head = ListNode(value[0])
            self.tail = self.head
            self.length = 1
            for element in value[1:]:
                self.append(element)
        else:
            self.head = ListNode(value)
            self.tail = self.head
            self.length = 1

    def append(self, value):
        new_node = ListNode(value)
        self.tail.next = new_node
        self.tail = new_node
        self.length += 1
        return self

    def prepend(self, value):
        new_node = ListNode(value)
        new_node.next = self.head
        self.head = new_node
        self.length += 1
        return self

    def lookup(self, value):
        return self.traverse_to_key(value)

    def traverse_to_key(self, key):
        current_node = self.head
        while current_node.next is not None:
            if current_node.val == key:
                return current_node
            current_node = current_node.next
        return False

    def remove_by_value(self, key):
        current_node = self.head
        previous_node = self.head
        while current_node.next is not None:
            if current_node.val == key:
                if current_node == previous_node:
                    self.head = current_node.next
                    self.length -= 1
                    return self.print_list()
                else:
                    previous_node.next = current_node.next
                    self.length -= 1
                    return self.print_list()
            previous_node = current_node
            current_node = current_node.next
        if current_node.next is not None and current_node.val == key:
            previous_node.next = None
            self.tail = previous_node
            self.length -= 1
            return self.print_list()
        return False

    def remove_by_index(self, index):
        if index > self.length:
            return False
        if index == 0:
            new_head_node = self.head.next
            self.head = new_head_node
            self.length -= 1
            return self.print_list()
        previous_node = self.traverse_to_index(index - 1)
        current_node = previous_node.next
        previous_node.next = current_node.next
        if previous_node.next is None:
            self.tail = previous_node
        self.length -= 1
        return self.print_list()

    def insert_by_index(self, index, value):
        if index >= self.length:
            self.append(value)
            return self.print_list()
        if index == 0:
            self.prepend(value)
            return self.print_list()
        new_node = ListNode(value)
        current_node = self.traverse_to_index(index)
        holding_pointer = current_node.next
        current_node.next = new_node
        new_node.next = holding_pointer
        self.length += 1
        return self.print_list()

    def insert_by_value(self, key, value):
        current_node = self.traverse_to_key(key)
        if current_node:
            if current_node.next is None:
                self.append(value)
                return self.print_list()
            new_node = ListNode(value)
            new_node.next = current_node.next
            current_node.next = new_node
            self.length += 1
            return self.print_list()
        else:
            self.prepend(value)
            return self.print_list()

    def print_list(self):
        current_node = self.head
        array = [current_node.val]
        while current_node.next is not None:
            current_node = current_node.next
            array.append(current_node.val)
        return array

    def traverse_to_index(self, index):
        counter = 0
        current_node = self.head
        while counter != index:
            current_node = current_node.next
            counter += 1
        return current_node

    def reverse(self):
        if self.length == 1:
            return self.print_list()
        current_node = self.head
        reversed_linked_list = (LinkedList(current_node.val))
        while current_node.next is not None:
            current_node = current_node.next
            reversed_linked_list.prepend(current_node.val)
        return reversed_linked_list.print_list()

    def reverse2(self):
        if self.length == 1:
            return self.print_list()
        first = self.head
        second = self.head.next
        self.tail = self.head
        while second:
            hold_node = second.next
            second.next = first
            first = second
            second = hold_node
        self.head.next = None
        self.head = first
        return self.print_list()


my_list = LinkedList(["val 1", "val 2", "val3", "other", "xxx", "val6"])
print(my_list.print_list())

my_list.append("new value")
print(my_list.print_list())

my_list.prepend("new first")
print(my_list.print_list())

my_list.remove_by_value("other")
print(my_list.print_list())

print(my_list.reverse())
print(my_list.traverse_to_index(2).val)
